awscli/customizations/logs.py

```diff
@@ -0,0 +1,321 @@
+        return is_a_tty()
+
+
+class TimestampUtils(object):
+    _RELATIVE_TIMESTAMP_REGEX = re.compile(
+        r"(?P<amount>\d+)(?P<unit>s|m|h|d|w)$"
+    )
+    _TO_SECONDS = {
+        's': 1,
+        'm': 60,
+        'h': 3600,
+        'd': 24 * 3600,
+        'w': 7 * 24 * 3600,
+    }
+
+    def __init__(self, now=None):
+        self._now = now
+        if now is None:
+            self._now = datetime.utcnow
+
+    def to_epoch_millis(self, timestamp):
+        re_match = self._RELATIVE_TIMESTAMP_REGEX.match(timestamp)
+        if re_match:
+            datetime_value = self._relative_timestamp_to_datetime(
+                int(re_match.group('amount')), re_match.group('unit')
+            )
+        else:
+            datetime_value = parse_timestamp(timestamp)
+        return int(datetime2timestamp(datetime_value) * 1000)
+
+    def to_datetime(self, millis_timestamp):
+        return datetime.fromtimestamp(millis_timestamp / 1000.0, tz.tzutc())
+
+    def _relative_timestamp_to_datetime(self, amount, unit):
+        multiplier = self._TO_SECONDS[unit]
+        return self._now() + timedelta(seconds=amount * multiplier * -1)
+
+
+class BaseLogEventsGenerator(object):
+    def __init__(self, client, timestamp_utils):
+        self._client = client
+        self._timestamp_utils = timestamp_utils
+
+    def iter_log_events(self, group_name, start=None, filter_pattern=None):
+        filter_logs_events_kwargs = self._get_filter_logs_events_kwargs(
+            group_name, start, filter_pattern)
+        log_events = self._filter_log_events(filter_logs_events_kwargs)
+        for log_event in log_events:
+            self._convert_event_timestamps(log_event)
+            yield log_event
+
+    def _filter_log_events(self, filter_logs_events_kwargs):
+        raise NotImplementedError('_filter_log_events()')
+
+    def _get_filter_logs_events_kwargs(self, group_name, start,
+                                       filter_pattern):
+        kwargs = {
+            'logGroupName': group_name,
+            'interleaved': True
+        }
+        if start is not None:
+            kwargs['startTime'] = self._timestamp_utils.to_epoch_millis(start)
+        if filter_pattern is not None:
+            kwargs['filterPattern'] = filter_pattern
+        return kwargs
+
+    def _convert_event_timestamps(self, event):
+        event['ingestionTime'] = self._timestamp_utils.to_datetime(
+            event['ingestionTime'])
+        event['timestamp'] = self._timestamp_utils.to_datetime(
+            event['timestamp'])
+
+
+class NoFollowLogEventsGenerator(BaseLogEventsGenerator):
+    def _filter_log_events(self, filter_logs_events_kwargs):
+        paginator = self._client.get_paginator('filter_log_events')
+        for page in paginator.paginate(**filter_logs_events_kwargs):
+            for log_event in page['events']:
+                yield log_event
+
+
+class FollowLogEventsGenerator(BaseLogEventsGenerator):
+    _TIME_TO_SLEEP = 1
```

Comment:
I would probably bump this up to at least 5 seconds.  We probably don't want to be running this command every second while we wait for more logs.

Reply:
Yeah I can bump that up.